src/ocl_util.cc

```diff
@@ -226,11 +246,18 @@ __constant uint32_t const U32_MAX = 0xffffffff;
       assert_st( !init_done.v );
       cl_get_platforms( platforms );
       if( platforms.empty() ) { rt_err( "no OpenCL platforms found" ); }
-      for( vect_cl_platform_id::const_iterator i = platforms.begin(); i != platforms.end(); ++i ) {
-	vect_cl_device_id devices;
-	cl_get_devices( devices, *i, CL_DEVICE_TYPE_GPU );
-	if( !devices.empty() ) { use_devices = vect_cl_device_id{devices[0]}; } // pick first device only (arbitrarily)
+      if( use_platform_id >= platforms.size() ) { rt_err( "desired platform ID is not available" ); }
+      vect_cl_device_id devices;
+      cl_get_devices( devices, platforms[use_platform_id], CL_DEVICE_TYPE_GPU );
+      if( !devices.empty() ) {
+	if( use_device_id >= devices.size() ) {
+          rt_err( "The specified device is not available withing in the specified platform" );
+        }
+	else {
+	  use_devices = vect_cl_device_id{devices[use_device_id]}; // pick the specified device
+        }
       }
+
       if( use_devices.empty() ) { rt_err( "no OpenCL platform had any GPUs (devices of type CL_DEVICE_TYPE_GPU)" ); }
```

Comment:
as per prior comment, this should probably be an 'assert size() == 1' now. use_devices doesn't really need to be a vector now-or-ever, but it's okay to stay one (i.e. one could plausible consider having multiple use_devices).